### PR TITLE
Lesson4

### PR DESCRIPTION
**Pull request by hlibb**

### CODE REVIEW

**Comment by hlibb** [comment]
3e5e493 is the last (but NOT FINAL!),  edition

File: src/lesson4/Calculator.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+    public void commandC() {
+        currentValue = 0;
+        calculatorLook(currentValue);
+    }
+
+    public void commandBackspace() { //only for currentValue
+        double number = currentValue;
+        String numberStr = Double.toString(number);
+        int i = 0;
+        while (numberStr.toCharArray()[numberStr.length() - 1] == '.' | numberStr.toCharArray()[numberStr.length() - 1] == '0') {
```
Comment (shaitanaleksander) [inline src/lesson4/Calculator.java:43]: complicated boolean operation, try to make it smaller or use different loop

File: src/lesson4/Calculator.java [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+            if (stringInCharArray[i] == ('1')
+                    | (stringInCharArray[i] == ('2')
+                    | stringInCharArray[i] == ('3')
+                    | stringInCharArray[i] == ('4')
+                    | stringInCharArray[i] == ('5')
+                    | stringInCharArray[i] == ('6')
+                    | stringInCharArray[i] == ('7')
+                    | stringInCharArray[i] == ('8')
+                    | stringInCharArray[i] == ('9')
+                    | stringInCharArray[i] == ('0'))) isDouble = true;
```
Comment (shaitanaleksander) [inline src/lesson4/Calculator.java:83]: if statement need to be short, you can create private function that iterate thought array and return boolean if the char is present in the array

File: src/lesson4/Calculator.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                } else System.err.println("Too many comas in your double value!");
+            }
+            else return false;
+        }
+        return isDouble;
+    }
+
+    public boolean isCommand(String command) {
+        return switch (command) {
+            case ("+"), ("-"), ("*"), ("/") -> true;
```
Comment (shaitanaleksander) [inline src/lesson4/Calculator.java:96]: looks bad, probably  creation array of all command will be good idea, you can iterate through it and return true if string is in the array and false if not, which  is bed without break at the end

File: src/lesson4/Calculator.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    public void useControlCommand(String controlCommand) {
+        switch (controlCommand) {
+            case ("c"): {
+                commandC();
+                break;
+            }
+            case ("<-"): {
+                commandBackspace();
+                break;
+            }
```
Comment (shaitanaleksander) [inline src/lesson4/Calculator.java:138]: if statement will be better

File: src/lesson4/Calculator.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        System.out.printf(
+                        " ----------------\n" +
+                        "|%15.2f |\n" +
+                        "| c   /   *   <- |\n" +
+                        "| 1   2   3   -  |\n" +
+                        "| 4   5   6   +  |\n" +
+                        "| 7   8   9      |\n" +
+                        "|     0   ,      |\n" +
+                        " ----------------\n"
+                , currentValue);
```
Comment (shaitanaleksander) [inline src/lesson4/Calculator.java:172]: this kind of string is better to save to variable

File: src/lesson4/Calculator.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    public void commandBackspace() { //only for currentValue
+        double number = currentValue;
+        String numberStr = Double.toString(number);
+        int i = 0;
+        while (numberStr.toCharArray()[numberStr.length() - 1] == '.' | numberStr.toCharArray()[numberStr.length() - 1] == '0') {
+            i++;
+            String numberStrEdited = numberStr.substring(0, numberStr.length() - i);
+            if ((numberStrEdited.toCharArray()[numberStrEdited.length() - 1] != ('.')) & (numberStrEdited.toCharArray()[numberStrEdited.length() - 1] != ('0'))) {
```
Comment (shaitanaleksander) [inline src/lesson4/Calculator.java:46]: same

File: src/lesson4/ArrayTask.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+        StringBuilder compliedWord = new StringBuilder();
+        for (String s : word) {
+            compliedWord.append(s);
+        }
+        return compliedWord.toString();
+    }
+
+    public boolean areEqual(String stringWord1, String stringWord2) {
+        return stringWord1.equals(stringWord2);
+    }
```
Comment (shaitanaleksander) [inline src/lesson4/ArrayTask.java:29]: you can remove it add compare two objects in launch() method

File: src/lesson4/ArrayTask.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    public void launch() {
+        System.out.println("Enter two strings to compare (use space to split the string):");
+//        System.out.println("Enter the first word array: ");  doesn't work after first loop, how to improve?
```
Comment (shaitanaleksander) [inline src/lesson4/ArrayTask.java:33]: do {} while  loop can help

Comment (hlibb) [reply src/lesson4/ArrayTask.java:33]: but then the 34 line doesn't work properly, such expression makes it impossible to quit during the first loop

File: src/lesson4/Calculator.java (hunk 1 above)
Comment (hlibb) [reply src/lesson4/Calculator.java:43]: i cant find out, how should i do this

**Comment by hlibb** [comment]
final

File: src/lesson4/ArrayTask.java (hunk 8 above)
Comment (shaitanaleksander) [reply src/lesson4/ArrayTask.java:29]: you can make addition check for exit command in the loop,  and if command is exit than do  return, and also do while {} can be with true statement,

File: src/lesson4/Calculator.java (hunk 1 above)
Comment (shaitanaleksander) [reply src/lesson4/Calculator.java:48]: you have int var I, why you don't try to use for() loop ? try to rewrite logic with while{} to for{} loop, if you did not succeed we can discus it on Friday

**Pull request closed without merging**